tests/rust/pipe-test/push_wait/client.rs

```diff
@@ -0,0 +1,120 @@
+        Ok(Self { libos, pipe_name })
+    }
+
+    // Runs the target pipe client.
+    pub fn run(&mut self) -> Result<()> {
+        let pipeqd: QDesc = self.libos.open_pipe(&format!("{}:rx", self.pipe_name))?;
+
+        // Allocate a scatter-gather array. Length and value are arbitrary.
+        let sga: demi_sgarray_t = match self.mksga(1, 123) {
+            Ok(sga) => sga,
+            Err(e) => {
+                // Clean up open pipe on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/638
+                anyhow::bail!("mksga() failed (error={:?})", e)
+            },
+        };
+
+        // Push scatter-gather array.
+        let qt: QToken = match self.libos.push(pipeqd, &sga) {
+            Ok(qt) => qt,
+            Err(e) => {
+                // Clean up open pipe on error.
+                // Clean up scatter-gather array on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/638
+                anyhow::bail!("push() failed (error={:?})", e)
+            },
+        };
+
+        // Succeed to release scatter-gather-array.
+        match self.libos.sgafree(sga) {
+            Ok(()) => {},
+            Err(e) => {
+                // Clean up open pipe on error.
+                // Clean up scatter-gather array on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/638
+                anyhow::bail!("sgafree() failed (error={:?})", e)
+            },
+        }
+
+        // Succeed to close pipe.
+        match self.libos.close(pipeqd) {
+            Ok(()) => {},
+            Err(e) => {
+                // Clean up open pipe on error.
+                // FIXME: https://github.com/demikernel/demikernel/issues/638
+                anyhow::bail!("close() failed (error={:?})", e)
+            },
+        }
+
+        // Wait for operation to complete.
+        match self.libos.wait(qt, Some(Duration::from_micros(0))) {
+            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_FAILED && qr.qr_ret == libc::ECANCELED => {},
+            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_PUSH && qr.qr_ret == 0 => {},
+            Ok(_) => anyhow::bail!("wait() should not complete successfully with an opcode other than DEMI_OPC_PUSH"),
+            Err(e) if e.errno == libc::ETIMEDOUT => anyhow::bail!("wait() should not timeout"),
+            Err(e) => anyhow::bail!("wait() should not fail (error={:?})", e),
+        }
+
+        Ok(())
+    }
+
+    /// Makes a scatter-gather array.
+    fn mksga(&mut self, size: usize, value: u8) -> Result<demi_sgarray_t> {
+        // Allocate scatter-gather array.
+        let sga: demi_sgarray_t = match self.libos.sgaalloc(size) {
+            Ok(sga) => sga,
+            Err(e) => anyhow::bail!("failed to allocate scatter-gather array: {:?}", e),
+        };
+
+        // Ensure that scatter-gather array has the requested size.
+        assert_eq!(sga.sga_segs[0].sgaseg_len as usize, size);
```

Comment:
Can you change this to demikernel::ensure_eq! ?

Reply:
Sure.